File: apps/rivet-team-android/app/src/main/java/dev/rivetos/team/gateway/StubGateway.kt
package dev.rivetos.team.gateway

import android.os.Handler
import android.os.Looper
import dev.rivetos.team.domain.LOCAL_USER_ID
import dev.rivetos.team.domain.Persona
import dev.rivetos.team.domain.SAMPLE_PERSONAS
import dev.rivetos.team.domain.TeamMessage
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

class StubGateway(
    override val baseUrl: String = "http://127.0.0.1:5174",
) : TeamGateway {
    private val threads = mutableMapOf<String, MutableList<TeamMessage>>()
    private val watchers = mutableMapOf<String, CopyOnWriteArrayList<(GatewayEvent) -> Unit>>()
    private val memory = mutableListOf<String>()
    private val main = Handler(Looper.getMainLooper())

    override fun listPersonas(userId: String): List<Persona> = SAMPLE_PERSONAS

    override fun sessionMessages(sessionId: String): List<TeamMessage> =
        threads[sessionId]?.toList() ?: emptyList()

    override fun postMessage(sessionId: String, text: String, userId: String) {
        val persona = SAMPLE_PERSONAS.firstOrNull { it.threadId == sessionId }
        val user = TeamMessage(
            id = UUID.randomUUID().toString(),
            sessionId = sessionId,
            userId = userId,
            personaId = persona?.id ?: sessionId,
            nodeId = persona?.nodeId ?: "local-node",
            role = "user",
            text = text,
            ts = System.currentTimeMillis(),
        )
        threads.getOrPut(sessionId) { mutableListOf() }.add(user)
        emit(sessionId, GatewayEvent.Message(user))
        emit(sessionId, GatewayEvent.Working("${persona?.name ?: "Persona"} is working…"))
        main.postDelayed({
            val reply = TeamMessage(
                id = UUID.randomUUID().toString(),
                sessionId = sessionId,
                userId = userId,
                personaId = persona?.id ?: sessionId,
                nodeId = persona?.nodeId ?: "local-node",
                role = "assistant",
                text = stubReply(persona?.name ?: "Persona", text),
                ts = System.currentTimeMillis(),
            )
            threads.getOrPut(sessionId) { mutableListOf() }.add(reply)
            memory.add("${persona?.name}: ${text.take(160)}")
            emit(sessionId, GatewayEvent.Done)
            emit(sessionId, GatewayEvent.Message(reply))
        }, 450)
    }

    override fun watch(sessionId: String, onEvent: (GatewayEvent) -> Unit): () -> Unit {
        val list = watchers.getOrPut(sessionId) { CopyOnWriteArrayList() }
        list.add(onEvent)
        return { list.remove(onEvent) }
    }

    override fun memorySearch(userId: String, q: String): List<String> {
        val needle = q.lowercase()
        if (needle.isBlank()) return emptyList()
        return memory.filter { it.lowercase().contains(needle) }
    }

    private fun emit(sessionId: String, event: GatewayEvent) {
        watchers[sessionId]?.forEach { it(event) }
    }

    private fun stubReply(name: String, userText: String): String {
        val clipped = userText.trim().replace(Regex("\\s+"), " ").take(220)
        return when (name) {
            "Summarizer" -> "Brief: $clipped\n\n• Point taken.\n• Next: wire the live gateway."
            "Informatics" -> "facts:\n- input: \"$clipped\"\n- status: captured (stub)"
            else -> "I would look into “$clipped”. Stub gateway — same contract as Hub, no live model yet."
        }
    }

    companion object {
        val shared = StubGateway()
        const val USER = LOCAL_USER_ID
    }
}

File: apps/rivet-team-web/src/lib/stub-gateway.ts
/**
 * In-process stub of the Rivet gateway session + memory surface.
 *
 * Same method names as RivetGateway. Replies are local so a reviewer can
 * send a message and see a working chip, then an assistant turn, without
 * a live node. Swap for `new RivetGateway({ baseUrl })` in a later slice.
 */

import type { TeamGateway } from './gateway.js'
import { appendMemory, searchMemory } from './memory.js'
import { personaByThread, SAMPLE_PERSONAS } from './seed.js'
import type {
  GatewayClientConfig,
  MemorySearchResponse,
  Persona,
  SessionMessage,
  SessionMessagesResponse,
  SessionPostAccepted,
  SessionPostRequest,
  SessionsListResponse,
  SessionWsFrame,
  Subscription,
  WatchOptions,
} from './types.js'
import { LOCAL_USER_ID } from './types.js'
import { uuidv4 } from './uuid.js'

interface Watcher {
  sessionId?: string
  onFrame: (frame: SessionWsFrame) => void
}

const threads = new Map<string, SessionMessage[]>()
const watchers = new Set<Watcher>()

function emit(frame: SessionWsFrame): void {
  const session = frame.kind === "message" ? frame.sessionId : frame.kind === "stream" ? frame.session : undefined
  for (const w of watchers) {
    if (w.sessionId && w.sessionId !== session) continue
    w.onFrame(frame)
  }
}

function stubReply(personaName: string, userText: string): string {
  const clipped = userText.trim().replace(/\s+/g, ' ').slice(0, 220)
  if (personaName === 'Summarizer') {
    return `Brief: ${clipped || '(empty)'}\n\n• Point taken.\n• Next: I would tighten this against source material once the live gateway is wired.`
  }
  if (personaName === 'Informatics') {
    return `facts:\n- input: "${clipped || ''}"\n- status: captured (stub)\n- next: persist via /api/sessions when the real RivetGateway is bound`
  }
  return `I would look into “${clipped || 'that'}”. This slice is the stub gateway — same postMessage / watchSessions contract as Hub, no live model yet.`
}

export function createStubGateway(config: GatewayClientConfig = { baseUrl: 'http://127.0.0.1:5174' }): TeamGateway {
  return {
    config,

    async listSessions(): Promise<SessionsListResponse> {
      return {
        sessions: SAMPLE_PERSONAS.map((p) => {
          const msgs = threads.get(p.threadId) ?? []
          return {
            id: p.threadId,
            lastActive: msgs.at(-1)?.ts ?? 0,
            messages: msgs.length,
          }
        }),
      }
    },

    async sessionMessages(sessionId: string): Promise<SessionMessagesResponse> {
      return { messages: [...(threads.get(sessionId) ?? [])] }
    },

    async postMessage(sessionId: string, body: SessionPostRequest): Promise<SessionPostAccepted> {
      const persona = personaByThread(sessionId)
      const userId = body.userId ?? LOCAL_USER_ID
      const userMsg: SessionMessage = {
        id: uuidv4(),
        sessionId,
        role: 'user',
        text: body.text,
        ts: Date.now(),
      }
      const list = threads.get(sessionId) ?? []
      list.push(userMsg)
      threads.set(sessionId, list)
      emit({ kind: 'message', ...userMsg })
      emit({
        kind: 'stream',
        session: sessionId,
        event: { type: 'status', content: 'working', metadata: { card: 'working' } },
      })

      const delay = 450
      setTimeout(() => {
        const reply: SessionMessage = {
          id: uuidv4(),
          sessionId,
          role: 'assistant',
          text: stubReply(persona?.name ?? 'Persona', body.text),
          ts: Date.now(),
        }
        const next = threads.get(sessionId) ?? []
        next.push(reply)
        threads.set(sessionId, next)
        emit({
          kind: 'stream',
          session: sessionId,
          event: { type: 'done', content: '' },
        })
        emit({ kind: 'message', ...reply })
        appendMemory({
          userId,
          content: `${persona?.name ?? 'persona'}: ${body.text.slice(0, 160)}`,
          role: 'user',
          agent: persona?.id ?? sessionId,
        })
      }, delay)

      return { accepted: true, session: sessionId }
    },

    watchSessions(onFrame, sessionId, opts: WatchOptions = {}): Subscription {
      const watcher: Watcher = { sessionId, onFrame }
      watchers.add(watcher)
      opts.onStatus?.('connecting')
      queueMicrotask(() => opts.onStatus?.('open'))
      return {
        close() {
          watchers.delete(watcher)
          opts.onStatus?.('closed')
        },
        send() {
          return true
        },
      }
    },

    async memorySearch(query): Promise<MemorySearchResponse> {
      return searchMemory(LOCAL_USER_ID, query)
    },

    async health(): Promise<boolean> {
      return true
    },

    listPersonas(_userId: string): Persona[] {
      return SAMPLE_PERSONAS
    },
  }
}
